packages/eslint-plugin/src/rules/expect.ts

```diff
@@ -0,0 +1,543 @@
+          context.report({
+            messageId,
+            data: { ...data, versionNameString: [...versions].sort().join(", ") },
+            loc,
+          });
+        }
+      },
+    };
+  },
+});
+
+const programCache = new WeakMap<ts.Program, Map<string, ts.Program>>();
+/** Maps a tslint Program to one created with the version specified in `options`. */
+function getProgram(configFile: string, ts: TSModule, versionName: string, lintProgram: ts.Program): ts.Program {
+  let versionToProgram = programCache.get(lintProgram);
+  if (versionToProgram === undefined) {
+    versionToProgram = new Map<string, ts.Program>();
+    programCache.set(lintProgram, versionToProgram);
+  }
+
+  let newProgram = versionToProgram.get(versionName);
+  if (newProgram === undefined) {
+    newProgram = createProgram(configFile, ts);
+    versionToProgram.set(versionName, newProgram);
+  }
+  return newProgram;
+}
+
+function createProgram(configFile: string, ts: TSModule): ts.Program {
+  const projectDirectory = path.dirname(configFile);
+  const { config } = ts.readConfigFile(configFile, ts.sys.readFile);
+  const parseConfigHost: ts.ParseConfigHost = {
+    fileExists: fs.existsSync,
+    readDirectory: ts.sys.readDirectory,
+    readFile: (file) => fs.readFileSync(file, "utf8"),
+    useCaseSensitiveFileNames: true,
+  };
+  const parsed = ts.parseJsonConfigFileContent(config, parseConfigHost, path.resolve(projectDirectory), {
+    noEmit: true,
+  });
+
+  if (config.compilerOptions?.module === "node16" && parsed.options.module === undefined) {
+    // TypeScript version is too old to handle the "node16" module option,
+    // but we can run tests falling back to commonjs/node.
+    parsed.options.module = ts.ModuleKind.CommonJS;
+    parsed.options.moduleResolution = ts.ModuleResolutionKind.NodeJs;
+  }
+
+  const host = ts.createCompilerHost(parsed.options, true);
+  return ts.createProgram(parsed.fileNames, parsed.options, host);
+}
+
+interface ReporterInfo {
+  versionName: string;
+  messageId: MessageIds;
+  data?: ReportDescriptorMessageData;
+  loc: Readonly<TSESTree.SourceLocation>;
+}
+
+type Reporter = (info: ReporterInfo) => void;
+
+const zeroSourceLocation: Readonly<TSESTree.SourceLocation> = {
+  start: { line: 0, column: 0 },
+  end: { line: 0, column: 0 },
+};
+
+function walk(
+  getLocFromIndex: (index: number) => Readonly<TSESTree.Position>,
+  report: Reporter,
+  fileName: string,
+  program: ts.Program,
+  ts: TSModule,
+  versionName: string,
+  nextHigherVersion: string | undefined,
+  dirPath: string,
+): void {
+  const sourceFile = program.getSourceFile(fileName)!;
+  if (!sourceFile) {
+    report({
+      versionName,
+      messageId: "programContents",
+      data: { fileName, versionName },
+      loc: zeroSourceLocation,
+    });
+    return;
+  }
+
+  const checker = program.getTypeChecker();
+  // Don't care about emit errors.
+  const diagnostics = ts.getPreEmitDiagnostics(program, sourceFile);
```

Comment:
I'll need to drop this call if we're using the `node_modules` copy of typescript; the editor will already show the errors and as of https://github.com/typescript-eslint/typescript-eslint/pull/7997 this is no longer safe.